### ISSUE TITLE
Data race in logger during testing

### ISSUE THREAD

**Issue by wxiaoguang** [go-gitea/gitea#24541]
```
==================
WARNING: DATA RACE
Read at 0x00c00b9398a3 by goroutine 144:
  testing.(*common).logDepth()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:992 +0xc4
  testing.(*common).log()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:985 +0x84
  testing.(*common).Log()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1026 +0x5b
  testing.(*T).Log()
      <autogenerated>:1 +0x55
  code.gitea.io/gitea/tests.(*testLoggerWriterCloser).Write()
      /home/runner/work/gitea/gitea/tests/testlogger.go:77 +0x28a
  code.gitea.io/gitea/modules/log.(*WriterLogger).LogEvent()
      /home/runner/work/gitea/gitea/modules/log/writer.go:234 +0x1ab
  code.gitea.io/gitea/tests.(*TestLogger).LogEvent()
      <autogenerated>:1 +0x44
  code.gitea.io/gitea/modules/log.(*ChannelledLog).Start()
      /home/runner/work/gitea/gitea/modules/log/event.go:85 +0x26a
  code.gitea.io/gitea/modules/log.NewChannelledLog.func1()
      /home/runner/work/gitea/gitea/modules/log/event.go:68 +0x39

Previous write at 0x00c00b9398a3 by goroutine 41423:
  testing.tRunner.func1()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1563 +0x82d
  runtime.deferreturn()
      /opt/hostedtoolcache/go/1.20.3/x64/src/runtime/panic.go:476 +0x32
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1629 +0x47

Goroutine 144 (running) created at:
  code.gitea.io/gitea/modules/log.NewChannelledLog()
      /home/runner/work/gitea/gitea/modules/log/event.go:68 +0x649
  code.gitea.io/gitea/modules/log.(*MultiChannelledLogger).SetLogger()
      /home/runner/work/gitea/gitea/modules/log/multichannel.go:33 +0x124
  code.gitea.io/gitea/modules/log.NewNamedLogger()
      /home/runner/work/gitea/gitea/modules/log/log.go:64 +0x204
  code.gitea.io/gitea/modules/log.NewLogger()
      /home/runner/work/gitea/gitea/modules/log/log.go:47 +0xad
  code.gitea.io/gitea/modules/setting.initLogFrom()
      /home/runner/work/gitea/gitea/modules/setting/log.go:325 +0x46b
  code.gitea.io/gitea/modules/setting.InitLogs()
      /home/runner/work/gitea/gitea/modules/setting/log.go:357 +0x58
  code.gitea.io/gitea/modules/setting.LoadSettings()
      /home/runner/work/gitea/gitea/modules/setting/setting.go:288 +0x8b
  code.gitea.io/gitea/routers.GlobalInitInstalled()
      /home/runner/work/gitea/gitea/routers/init.go:123 +0x718
  code.gitea.io/gitea/tests.InitTest()
      /home/runner/work/gitea/gitea/tests/test_utils.go:175 +0x1fe7
  code.gitea.io/gitea/tests/integration.TestMain()
      /home/runner/work/gitea/gitea/tests/integration/integration_test.go:88 +0xc4
  main.main()
      _testmain.go:959 +0x324

Goroutine 41423 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1629 +0x805
  code.gitea.io/gitea/tests/integration.TestPackageContainer.func5()
      /home/runner/work/gitea/gitea/tests/integration/api_packages_container_test.go:178 +0x825
  testing.tRunner()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /opt/hostedtoolcache/go/1.20.3/x64/src/testing/testing.go:1629 +0x47
```

**Comment by silverwind** [comment]
After fix, we should probably make these errors fatal.

**Comment by wxiaoguang** [comment]
It is fatal ....

**Comment by silverwind** [comment]
Oh, ok. I wasn't sure because of `WARNING`, but if this exits the test non-zero, it's fine.

**Comment by wxiaoguang** [comment]
According to Go:

```
		// Do not lock t.done to allow race detector to detect race in case
		// the user does not appropriately synchronize a goroutine.
		t.done = true
```

Gitea doesn't follow this constraint

**Comment by silverwind** [comment]
> The problem is that our logger system tries to output logs when the runner is finishing.

I wonder if we can run integration tests through standard `go test` instead of this homebrew test runner we have. Also see https://github.com/go-gitea/gitea/pull/24528#issuecomment-1535854727 about it.

**Comment by wxiaoguang** [comment]
Gitea is using standard `go test`

However, the problem is the legacy Gitea's `log` package, it has too many technical debts, and it should be rewritten.

**Comment by silverwind** [comment]
> Gitea is using standard go test

Are you sure? Could the logger just be removed and it would output regular `go test` output? https://github.com/go-gitea/gitea/pull/24528 was made specifically to make the integration tests more readable but than I ran into this logger problem.

**Comment by wxiaoguang** [comment]
The problem is that a lot of Gitea's code is using its own `modules/log` package.

So `testlogger` is used to bridge the logs from `modules/log` package to `testing` package.

**Comment by wxiaoguang** [comment]
Hmm ... I think I know what do you mean by "standard" now .....

Well, Gitea uses `PrintCurrentTest` in `testlogger.go` to print test results, that's not standard .....

**Comment by silverwind** [comment]
It's far from standard. The integration tests invocate with

```
GITEA_ROOT="/home/runner/work/gitea/gitea" GITEA_CONF=tests/sqlite.ini ./migrations.sqlite.test
```

Where `migrations.sqlite.test` is a gitea binary seemingly build specificially for the integration test. A better approach would be to run the normal binary as a actions `service` and then run `go test` to preform the actual testing. No logger needed with that approach.

**Comment by wxiaoguang** [comment]
`migrations.sqlite.test` is standard, it's generated by `go test -o`, that's Golang recommended method: pre-build a binary for multiple executions.

```
go help test
	-o file
	    Compile the test binary to the named file.
	    The test still runs (unless -c or -i is specified).
```

**Comment by silverwind** [comment]
I wonder if I can get `gotestsum` to use these binaries while disabling the logger module. I would assume no because it wraps `go test`, not these output binaries.

**Comment by wxiaoguang** [comment]
The answer could only be clarified by reading source code.

If we are not in a hurry, we can postpone it to after the logger refactoring - which is also in my plan.

**Comment by wxiaoguang** [comment]
The data race bug could be fixed by #24505 919f157849b4828d684901395d6cd346a340d4bf , which doesn't occur with old queue (because old queue's flush is slow, so this bug was never triggered).

So close this one.
